Add 'WeightedPSQ' and use it to sort package, flag, and stanza choices.

grayjay opened the pull request:
I split this out from #2917 because the data structure is also needed for goal-scoring (#3488).  `WeightedPSQ` associates weights with choices in the search tree.  This PR uses `WeightedPSQ` for package, flag, and stanza choices and `PSQ` for goal choices, but eventually all choices should use `WeightedPSQ`.  There are a few TODOs in the code.

/cc @kosmikus 

EDIT: This PR uses `[Double]` as the weight type to avoid changing the behavior of the solver.

ezyang commented:
Let me be the grumpy code reviewer:

1. Needs docs!
2. Needs asymptotics!
3. Degree is in an odd place

But I'm all for defining data structures we need.

kosmikus commented:
@grayjay Looks promising so far. It doesn't seem to have any negative performance impact afaics.

grayjay commented:
Thanks.  I moved `Degree` into its own module, and I'll add documentation and unit tests next.

@kosmikus How do you want to proceed?  Should I open a new PR with just `WeightedPSQ` so that it is easier to merge, or should I continue cleaning up all of these changes?  I realized that I can remove some of the TODOs because they only relate to install plan scoring.

grayjay commented:
I cleaned it up, added tests, and added documentation with asymptotics, similar to D.Compat.Graph.  It's ready for review.  I'd like to clean up the history before it is merged, though.

ezyang commented on cabal-install/Distribution/Solver/Modular/WeightedPSQ.hs:
```diff
@@ -0,0 +1,102 @@
+filter :: (v -> Bool) -> WeightedPSQ k w v -> WeightedPSQ k w v
+filter p (WeightedPSQ xs) = WeightedPSQ (L.filter (p . triple_3) xs)
+
+-- | /O(N)/.
```
Ouch! Should there be a TODO here to track the length separately?

grayjay replied:
`length` is actually no longer called.  I think it was used by --reorder-goals, which now uses `degree` to avoid traversing the whole list.  Calling `length` there had a significant impact on performance because it was called on the children of every available goal choice after the children were filtered with an expensive predicate.  I'll remove it.

ezyang replied:
Cool, thanks.

23Skidoo commented:
@grayjay @kosmikus Is this ready to go in?

grayjay commented:
I was done working on it, but I wasn't sure if @kosmikus wanted to take another look.

23Skidoo commented:
Let's merge then, if @kosmikus dislikes something he can always revert.
